README.md
# <p align="center"> MKlon3.5v7
## <p align="center">Зарядное устройство на модулях ESP32DEVKITCV4, SAMD21 MINI и ILI9486 3.5".
### <p align="center">(реализация конечного автомата в проекте зарядного устройства)
## <p align="center">Проект Arduino.
#### <p align="right">2023.11.09

### MKlon3.5 - это вариант компоновки проекта MKlon2 с сохранением основных параметров, исключая разве что пользовательский интерфейс, реализованный на дисплее ILI9486 3.5 дюйма с сенсорной панелью. Реализована "этажерочная"  конструкция из трёх плат: дисплей и две платы размером 100х65 мм — плата управления и силовая плата. Предусмотрены разъемы для подключения логического анализатора к выводам обоих модулей - ESP32 и SAMD21 и сигнальные светодиоды.

### Аппаратная часть проектировалась в среде Eagle v7.3.0, программная - в PlatformIO. 

<a name="menu"></a>
1. [Как это работает. Аппаратный аспект.](#hard)
- [Межпроцессорное соединение.](#cpu)
- [Межпроцессорный обмен.](#uart)
- [Измерители.](#measure)
- [Схема разряда.](#discharge)
- [Подключение заряжаемой батареи.](#akb)
- [Питание модулей.](#dc5v)
- [Силовой DC/DC преобразователь.](#power)
- [Подключение вентилятора.](#cooler)
- [Конструктивное исполнение.](#design)
2. [Как это работает. Программный аспект](#soft)
- [Структура проекта](#main)
- [State.](#state)
- [Вычисление среднего](#sum)
3. [Технические характеристики](#specifications)
4. [FSM](#fsm)
5. [Первый проект](#first_project)
6. [Драйвер SAMD21](#driver)
- [Выбор аналоговых портов](#port)
- [Коэффициент пересчета в миллиамперы](#kma)
- [Коэффициент пересчета в милливольтры](#kmv)
- [Выбор таймера ШИМ](#timer2")
- [Выбор ПИД-регулятора](#fast_pid)
- [Подбор параметров](#par_pid)
7. [Документы](#docs)
- [Схема платы управления](#sch1)
- [Схема силовой платы](#sch2)
8. [Полезные ссылки](#useful_links)
9. [About Me](#about)

***
![](https://github.com/olmoro/MKlon3.5v7/blob/main/documents/full/img/MKlon3.5full.png)

## <p align="center"> Как это работает.<a name="hard"></a>
В зарядном устройстве пользовательский интерфейс реализован на модуле с ESP32 (ESP), а управление модулем силовой платы производится по последовательному каналу. На ESP обслуживается дисплей, RGB индикатор, зуммер, система контроля питания, система охлаждения, интерфейс драйвера силовой платы, USB, WiFi, BT. В памяти ESP хранятся заводские и пользовательские настройки. Оператору предоставляется возможность выбрать посредством меню требуемый режим работы, задать параметры и запустить. Настройки беспроводной сети позволяют в реальном времени на удаленном компьютере не только наблюдать в виде графиков процесс: напряжение, ток, баланс заряда/разряда, температуру, работу системы охлаждения, но и управлять им. Программное обеспечение ESP построено таким образом, что позволяет даже непрофессиональному программисту реализовать свой собственный алгоритм, воспользовавшись методикой из репозитория FSM.

Весь, или почти весь реалтайм реализован на отдельном модуле SAMD21 MINI (D21). Связь c ESP осуществляется по асинхронному каналу. Модуль D21 получает команду, по которой включаются соответствующие ресурсы: АЦП тока и напряжения, ШИМ-генератор, ЦАП управления разрядом, коммутатор выхода и др. Автоматически поддерживается устойчивость силового преобразователя при работе на малую нагрузку или на холостом ходе, отключение при перегрузках. Поддерживается режим быстрого разряда накопительных конденсаторов через нагрузку для разряда, минуя датчик тока, дабы не искажать подсчет ампер-часов. Когда возможно допускается регулировка параметров "на лету", иначе D21 переводит силовую часть в безопасный режим с выдачей сообщения управляющему контроллеру. D21 поддерживает управление зарядом и разрядом как командами, так и с помощью собственного ПИД-регулятора. Настройки коэффициентов для поддержания тока и напряжения могут быть разные. Выбор ПИД-регулирования имеет 4 режима: отключено, по току, по напряжению и с автовыбором по типу реализованного в TL494. Измерения напряжения на батарее производятся по четырехпроводной схеме в диапазоне от -2 до +18 вольт. Ток разряда измеряется на том же шунте, что и ток заряда от -10 до +10 ампер. Параметры фильтрации измерений задаются командами.

Питание прибора производится от внешнего AC/DC преобразователя (БП) на 18...19 вольт 6...9 ампер, что определяет выходные параметры зарядного устройства и делает само устройство с точки зрения электробезопасности более привлекательным.

В разделе "Как это работает" позиционные обозначения на рисунках могут отличаться от имеющихся в документации.

***
## <p align="center">1. Как это работает. Межпроцессорное соединение.<a name="cpu"></a>

![](https://github.com/olmoro/MKlon3.5v7/blob/main/documents/full/img/interprocessor.png)

Здесь реализована как минимум одна хитрость: модуль SAMD21 MINI запитан от изолированного источника питания со смещением в "минус" примерно на 200 милливольт, необходимые для дифференциальных измерений напряжения и тока. При реализованной топологии печатной платы существенного влияния на обмен по асинхронному интерфейсу не обнаружено.

Однако есть и недостаток: при одновременном подключении к одному компьютеру обоих USB указанное смещение становится равным нулю. И в результате как напряжение, так и ток отрицательной полярности при любом их значении индицируются вблизи нуля. Всё бы ничего, но в режиме разряда PID-регулятор поддержания установленного тока просто "сходит с ума" - всё добавляет и добавляет, пока ток не "упрется" в максимум, определённый сопротивлением нагрузки.

Но до этого, думаю, дело не дойдёт - это ж надо быть каким крутым программистом, чтобы одновременно кодировать два разных проекта...

Активным в связке является ESP32, который генерирует запрос по асинхронному интерфейсу. Линия READY - резервная, При экспериментах служит для синхронизации логического анализатора.

В начало [^](#menu)

***
## <p align="center"> Как это работает. Межпроцессорный обмен.<a name="uart"></a>

  Связь между управляющим и измерительным контроллерами в части физического интерфейса 
состоялась в пользу UART, аппаратная реализация которого намного проще, чем прожорливый 
и склонный к "неожиданностям" капризный I2C.

  На логическом уровне обмена данными существуют различные протоколы, но они обычно 
избыточны для реализации управления простыми микроконтроллерными устройствами, а потому 
применять такие протоколы не имеет смысла и выбор пал на протокол Wake, который был создан 
специально для реализации такой связи.

  Для надежного обнаружения начала пакета в потоке данных протокол Wake использует 
специальный зарезервированный символ, который в потоке данных встречаться не может. При 
необходимости передачи такого символа он заменяется последовательностью из двух символов 
(используется так называемый байт-стаффинг).

  Обмен идет посылками на скорости 230400 бод. Формат посылки (команды), как запроса, 
так и ответа выбран таким, чтобы укладываться в отведенные 500-600 микросекунд между 
регулярными циклами (1кГц) измерений тока и напряжения, включающих регулирование, 
обработку перегрузок и т.п. В пересчете на байты это не более 6-8 байт данных плюс 4 
служебных (стартовый, код команды, число байт и контрольная сумма пакета) и ещё должен 
быть запас на возможный байт-стаффинг. Данные передаются только как целочисленные и 
только один раз за 100 мс.

  Управляющий ESP32, если нет запроса на посылку целевой команды, отправляет запрос на 
получение данных от измерителя (2 байта напряжения и 2 байта тока) и 2 байта состояния 
SAMD21. Напряжение и ток в физических единицах - милливольтах и миллиамперах со знаком.
А два байта состояния дают исчерпывающие данные для управления процессом, учитывая то, 
что весь "реал-тайм" исполняется контроллером SAMD21 и не требует "мгновенного" 
вмешательства управляющего ESP32 в процесс целевой командой. В таком случае одно 
измерение из 100 заменяется предыдущим, что практически не заметно при визуализации как 
на дисплее, так и браузере.

В начало [^](#menu)

***
## <p align="center"> Как это работает. Измерители.<a name="measure"></a>

![](https://github.com/olmoro/MKlon3.5v7/blob/main/documents/full/img/ui_measure.png)

Измерение напряжения и тока производится по дифференциальной схеме. Микроконтроллер SAMD21 предоставляет такую возможность. Только следует учесть, что 12-разрядный АЦП становится фактически 11-разрядным плюс знаковый разряд. И рассчитывая параметры шунта и делителя напряжения следует оперировать половинным значением опорного напряжения. На входах установлены НЧ фильтры и диоды защиты. А чтобы при манипулировании с "крокодилами" вход измерителя напряжения не повисал неподключенным, одноимённые силовой и измерительный провод соединены резисторами (на рисунке не показаны).

В начало [^](#menu)

***
## <p align="center"> Как это работает. Вычисление среднего.<a name="sum"></a>

Не вызывает никаких сомнений метод вычисления среднего за определенный интервал, когда сумму измерений делят на их число. Однако для получения следующего среднего придется всё повторить с начала, и на это потребуется время. Но есть метод, когда на следующем цикле измерения убирается (вычитается) самый ранний уровень и добавляется (прибавляется) новый, а затем вычисляется среднее. Всё бы хорошо, но чтобы сохранить все эти уровни может потребоваться немалый расход памяти. Есть метод, позволяющий устранить этот недостаток, заменив вычитаемый ранний уровень средним значением, и называется он скользящим средним. Наибольший выигрыш дает не произвольное число измерений, а кратное степеням двойки: 2, 4, 8, 16, тогда операцию деления можно заменить арифметическим сдвигом. Соответственно буфер (sum) для накопления должен иметь такое число разрядов, чтобы не возникало переполнения. Функция вычисления для 16-разрядного ADC может выглядеть так:
```c++
int16_t averageU(uint8_t k) 
{
  static int32_t sum = 0;  // объявили и инициализировали сумматор удвоенной, чем ADC разрядности 
  int16_t avrU = sum >> k;  // вычислили среднее за прошлый интервал (накопленное поделили на число измерений)
  sum += adcU - avrU;    // к сумматору добавили разность между новым измерением и средним
  return int16_t(sum >> k);  // вернули новое значение среднего
}
```
Внимание! Будьте внимательны со скобками - в C++ приоритеты примененных здесь операций в порядке убывания: ()  -  >>  +=  .
То есть среднее равно (sum += adcU - (sum >> k)) >> k  - это без округления, младшие биты просто отбрасываем. С округлением к результату следует прибавить:
```c++
  ((sum >> (k - 1)) & 1);
```
В этом проекте одновременно с подсчетом среднего производится преобразование 12-разрядного ADC в 16-разрядный с помощью простого приема, для этого при вычислении среднего за прошлый интервал результат сдвигается на k плюс число недостающих разрядов ADC, то есть на k + 4. Если бы не одно "но". Когда колебания измеряемого параметра укладываются в единицу счета АЦП, то и результат будет стремиться к "родному" 11-разрядному. С измерением напряжения на батарее скорее всего так и будет (плюс-минус 10 милливольт), а вот с током заряда, где колебания обусловлены физико-химическими процессами, плюс-минус 1 миллиампер вполне достижим. 

А есть ли у метода "скользящее среднее" недостаток? Есть, в начальный период, пока сумматор не накопит достаточного количества измерений, значение среднего может оказаться существенно меньшим, чем реальное значение. Но и здесь не всё так печально. Во-первых, это можно использовать в мирных целях, например для "мягкого старта" системы регулирования. А во-вторых, объявляя сумматор инициализировать его не нулем, а некоторой величиной, например при измерении температуры радиатора задать величину, соответствующую температуре окружающей среды, не забыв сдвинуть влево на "k" разрядов. Для измерителя напряжения ЗУ это будет около 12 вольт. Так сумматор быстрее наберёт необходимое число измерений.

В начало [^](#menu)
***
## <p align="center"> Как это работает. Схема разряда.<a name="discharge"></a>

![](https://github.com/olmoro/MKlon3.5v7/blob/main/documents/full/img/i_discharge.png)

Измерение тока разряда производится на том же шунте RS, что и ток заряда, исключая неоднозначность при подсчете залитых и слитых ампер-часов. Для облегчения теплового режима предусмотрен разъем для подключения внешней нагрузки. В случае отсутствия таковой ток ограничивается двух-ваттным резистором. В режиме заряда при чрезмерно малой нагрузке силовой DC/DC автоматически подгружается этим резистором для обеспечения его устойчивой работы, причем этот ток не фиксируется шунтом, а потому не вносит погрешности в подсчет ампер-часов.

В начало [^](#menu)
***
## <p align="center"> Как это работает. Подключение заряжаемой батареи.<a name="akb"></a>

![](https://github.com/olmoro/MKlon3.5v7/blob/main/documents/full/img/switch.png)

Здесь ничего нового - схема "цельнотянутая" с прототипа, коим является Кулон 912 и 920-й. Даже предохранитель и 300-амперный диод защиты от повреждения электролитических конденсаторов силового преобразователя при переполюсовке вписались как родные. Но это не более чем перестраховка - при разработке всякое может случиться. Последний барьер не помешает. В качестве ключей применены HEXFET МОП-транзисторы IRLR2905 с высокой скоростью переключения и повышенной лавиноустойчивостью, а также имеющие возможность управляться логическим уровнем. То есть оба изолированных источника в приборе могут быть 5-вольтовые. Посмотрим, оправдается ли такой выбор. Управление ключами производится через оптопару.

В начало [^](#menu)
***
## <p align="center"> Как это работает. Питание модулей.<a name="dc5v"></a>

![](https://github.com/olmoro/MKlon3.5v7/blob/main/documents/full/img/power.png)

Питание модулей и схемы подключения батареи, коей нужна "дежурка" с гальваноразвязкой. Оказалось, что это достаточно просто получить, используя 1-ваттный преобразователь из 5 вольт входных в 12 вольт выходных. Аналогичный DC/DC на 5 вольт выходных используется для питания измерительного модуля SAMD21 для сдвига на 200 милливольт в минус от общего провода силового ИП. Эти пребразователи не имеют встроенного стабилизатора, но в обоих случаях это и не требуется. Подключение же их произведено со всеми рекомендациями - самовосстанавливающийся предохранитель на входе, LC-фильтры по входу и выходу, резисторы нагрузки и конденсатор между предполагаемыми "холодными" концами обмоток. 

VD6 следует выбирать с минимальным обратным током, в лучшем случае он добавит пару милливольт к разности напряжений на клеммах.

В начало [^](#menu)
***
## <p align="center"> Как это работает. Силовой DC/DC преобразователь.<a name="power"></a>

![](https://github.com/olmoro/MKlon3.5v7/blob/main/documents/full/img/dc_dc.png)

Выбрана схема преобразователя с P-MOSFET IRFP9540 и драйвером MIC4420 как вполне достаточная, чтобы получить возможность отладки алгоритмов управления, оставляя на будущее масштабирование вольтамперных  характеристик. В качестве "донора" был выбран [XL4016 DC-DC Max 9A 300W](https://aliexpress.ru/item/32661166685.html?_ga=2.264151500.1168692661.1643047564-769654542.1642920280&sku_id=12000020944321893&spm=a2g39.orderlist.0.0.60754aa6RxSv8u), силовые компоненты которого, кроме 4016, были перенесены на плату прибора. Рабочая частота была воспроизведена путем настройки режима работы таймера SAMD21. Несмотря на относительно высокую частоту микроконтроллера (48МГЦ) и турбо-режим (72МГц) удалось получить лишь 9-разрядный выход 190-килогерцового ШИМа. Вкупе с ПИД-регулятором это обеспечивает заявленные дискретности установки тока и напряжения 10 мА и 10 мВ.

О резисторе на входе драйвера. Выяснилось, что в процессе работы может последовать включение питания без модуля SAMD21, и, чтобы избежать неконтролируемого напряжения на выходе преобразователя, пришлось подтянуть оказавшийся "в воздухе" вход к плюсу питания. Естественно, что если применен инвертирующий драйвер типа MIC4429, то вход надо подтянуть к общему проводу.

Выбор резистора в цепи затвора определяется следующими требованиями:
1. ограничение тока драйвера (ниже допустимого для драйвера) при переключениях;
2. ограничение скорости изменения напряжения на стоке, дабы не привести к броскам напряжения на затворе через емкость Миллера с возможным повреждением затвора и драйвера высоким напряжением;
3. ограничение добротности затворной цепи, то есть колебательного контура, образованного емкостью затвора и паразитными индуктивностями проводников между драйвером затвора и самим кристаллом МОСФЕТа. Ибо высокодобротная цепь приводит к резонансным выбросам на затворе и возможности повреждения высоким напряжением как самого затвора, так и драйвера.

В итоге из доступных вариантов HCPL3120, UCC37322, транзисторный или MIC4420 предпочтительным по параметрам оказался последний. Опторазвязка, встроенные цепи компенсации эффекта Миллера или дешевизна реализации против 20-вольтовых MIC в схеме с конструктивно обеспеченном диапазоне напряжения питания 19-ю вольтами и выбором MOSFET с предельным напряжением исток-затвор 20 вольт дают возможность до предела упростить схему и получить хорошую повторяемость при использовании источника питания с гарантированными выходными параметрами.

В начало [^](#menu)

***
## <p align="center"> Как это работает. Подключение вентилятора.<a name="cooler"></a>

![](https://github.com/olmoro/MKlon3.5v7/blob/main/documents/full/img/cooler.png)

Система управления охлаждением охвачена обратной связью через датчик температуры типа NTC, установленный на радиаторе. Напряжение, подаваемое на вентилятор, управляется ПИД-регулятором. Порог температуры и коэффициенты Kp, Ki и Kd задаются настройками: 
![](https://github.com/olmoro/MKlon3.5v7/blob/main/documents/full/img/PID_Compensation_Animated.gif)

Как видно из иллюстрации, можно задать как ленивый, так и агрессивный характер системы. Мне понравился агрессивный, он же с перерегулированием - как только температура превысит порог, быстро её "сдуть" типа как бы чего не вышло, и через несколько таких включений-выключений установится некая средняя скорость вентилятора для поддержания температуры на заданном уровне.

В начало [^](#menu)

***
## <p align="center"> Конструктивное исполнение.<a name="design"></a>

![](https://github.com/olmoro/MKlon3.5v7/blob/main/documents/full/img/design.png)

Конструктивно зарядное устройство выполнено на трёх печатных платах - покупной дисплей и две заказные: плата управления и силовая плата. 

- [дисплей ILI9486 480*320 ](https://aliexpress.ru/item/1005001999296476.html?spm=a2g2w.orderdetail.0.0.76554aa69gSMCe&sku_id=12000018365356569) 3.5 дюйма с межосевым расстоянием между разъёмами (по факту) 92.5 мм. Поставщик хорошо упаковывает товар, тем и определился выбор. Сенсорная панель и карта памяти подключены.

- плата управления (cpu), гербер-файл cpu3.5v5.zip, имеет размер 99.7 * 61.6 мм, что позволяет заказать изготовление с доставкой пяти штук за эквивалент $10.2 - около 800р.

- силовая плата (pow), гербер-файл pow3.5v5.zip, имеет тот же размер за такие же деньги.

Механическое крепление - пластиковые стойки 12 мм для дисплея и 20 мм для силовой платы. Электрическое соединение плат cpu и pow выполнено на [удлиннённых штыревых разъёмах](https://aliexpress.ru/item/32995210479.html?spm=a2g2w.orderdetail.0.0.710a4aa675RwIh&sku_id=66971004869) с шагом 2.54 мм, что не исключает использование шлейфа в процессе отладки.

На модули [ESP32_DEVKITC_V4](https://aliexpress.ru/item/1005002443255665.html?spm=a2g2w.cart.cart_split.2.27194aa6uI5WfP&sku_id=12000020676652656&_ga=2.47565607.1828311252.1683559036-197161830.1681374836) и [SAMD21 M0-Mini](https://aliexpress.ru/item/1005002585114846.html?sku_id=12000021271636987&spm=a2g2w.productlist.search_results.0.58ef4aa657Lxkj) устанавливаются [штыри](https://aliexpress.ru/item/4000242164907.html?sku_id=10000000983312027&spm=a2g2w.productlist.search_results.184.5cd74aa6nJZy0m), а на плату соответственно [цанговые гнёзда](https://aliexpress.ru/item/4000131109405.html?sku_id=10000000387688826&spm=.search_results.0.140e4aa6Y83Jna) - они и надёжнее, и позволят модулям поместиться между платами прибора.

Почему 38-пиновая ESP32_Devkitc_V4, ведь на ней установлены только элементы питания и интерфейса? Вот и хорошо, в соответствии с аппетитом установим [ESP32-WROOM-32D](https://aliexpress.ru/item/1005002032158386.html?sku_id=12000022537417357&spm=a2g2w.productlist.search_results.5.71974aa6mf3A2W) с 8 или 16 МБ вместо обычных 4 МБ да и габарит подходящий для установки на 62-х миллиметровую плату - и USB доступен, и антенна экранируется платами в меньшей степени. А припаять, поверьте, гораздо проще, чем выпаивать для замены. И штыри паять только после распайки ESP32.  


history.txt
.LOG
22:00 25.04.2023

раскомментированны следующие строки:

// #define ILI9341_DRIVER
#define ILI9486_DRIVER


#define TFT_HEIGHT 320 // ST7789 240 x 320

#define TFT_INVERSION_OFF
                 My:
#define TFT_MISO 19
#define TFT_MOSI 23
#define TFT_SCLK 18
#define TFT_CS   32      // Chip select control pin
#define TFT_DC   05      // Data Command control pin
#define TFT_RST  04      // Reset pin (could connect to RST pin)

#define TOUCH_CS 33      // Chip select pin (T_CS) of touch screen
//            BL +3.3    // 21 (for v7)


#
#define SMOOTH_FONT

#define SPI_FREQUENCY  27000000
#define SPI_READ_FREQUENCY  20000000
#define SPI_TOUCH_FREQUENCY  2500000

===================
  TFT_BLACK
  TFT_BROWN
  TFT_RED
  TFT_ORANGE
  TFT_YELLOW
  TFT_GREEN
  TFT_BLUE
  TFT_PURPLE
  TFT_DARKGREY
  TFT_WHITE
  TFT_CYAN
  TFT_MAGENTA
  TFT_MAROON
  TFT_DARKGREEN
  TFT_NAVY
  TFT_PINK
  Если цвета на вашем дисплеи получаются инвертированными (например: черный цвет 
  отображается как белый), то расскоментируйте 
  #define TFT_INVERSION_OFF и закоментируйте 
  #define TFT_INVERSION_ON или наоборот в файле User_Setup.h
==================
Вывод текста
drawString(Текст, x, y);
tft.drawString("Hello Blog", 10, 10);

Рисование точки
drawPixel(x, y, цвет);
tft.drawPixel(10, 10, TFT_RED);

Рисование круга
drawCircle(x, y, радиус, цвет); - только границы
fillCircle(x, y, радиус, цвет); - заполненный круг
tft.drawCircle(100, 100, 40, TFT_BLUE);

Рисование прямоугольника
drawRect(x1, y1, ширина, высота, цвет); - только границы
fillRect(x1, y1, ширина, высота, цвет); - заполненный прямоугольник
tft.drawRect(10, 10, 100, 45, TFT_MAGENTA);

Рисование треугольника
drawTriangle(x1, y1, x2, y2, x3, y3, цвет); - только границы
fillTriangle(x1, y1, x2, y2, x3, y3, цвет); - заполненный треугольник
tft.drawTriangle(10, 160, 90, 10, 170, 160, TFT_PINK);

Рисование изображения
Рассмотрим пример, где у нас есть готовый файл и изображением.
Нужно конвертировать изображение в XBM формат, можно воспользоваться онлайн конвертерами.
Теперь нужно открыть XBM файл спомощю текстового редактора, скопировать содержимое в свой скетч.
drawXBitmap(x, y, данные изображения, ширина изображения, высота изображения, цвет изображения,
цвет фона);

22:12 26.04.2023
Первое включение дисплея ILI9486 3.5"    			    - OK

21:18 03.05.2023
Первое включение от +5 и +12.5, измерения V, F, T - OK

08:00 04.05.2023
mpins.h: распиновка платы с модулем ESP v7        - OK
Включение от +19.6 0.03А (без WiFi)               - OK
С вентилятором (DC 24V) 100% 18.4в 0.05А          - OK

08:00 06.05.2023
SDO дисплея к V_MISO не подключать!!!! исправлено - OK

09:00 08.05.2023
Проверена совместимость с Espressif 32 6.3.1 (без connectTask) - NO
WiFi 192.168.0.12                                 - OK

19:20 09.05.2023
Новый класс MTouch (MKeyboard пока оставлен)      - OK

09:00 10.05.2023
MDisplay: Проверено время выдачи обязательных данных 30мс, поделены на 
    printVolt()                               // 11 ms
    printAmp()                                // 11 ms
    printAmpHour(ah, 1)                       //  2 ms
    printCelsius(cels, 1)                     //  2 ms
    printDuration(duration, 0) + printHeap()  //  3 + 3 ms
Обработка Touch с getTouchRaw(&x, &y) производится каждые 25 мс
При вызове задачи через 25 мс время полного обновления
не изменилось 25мс * 10 = 250мс                                 - OK
Класс MTouch возможно не будет использоваться.

18:00 24.06.2023
  В задаче measureTask оставлены только измерения, проиводимые ресурсами ESP32:
температуры и напряжение источника питания. Аппаратных кнопок в приборе нет.


7:21 30.06.2023
  После очистки от неиспользуемых старых файлов проекту дано новое название
MKlon3.5v7 по версии корректированных плат.

11:00 09.07.2023
  Удалены неиспользуемые файлы.
  Удален BUTTON.
  Вызовы Keyboard->getKey() частично заменены на Display->getKey() или закомментированы.
  Архив  MKlon3.5v7_20230709-1.zip

=============================================
  Обработчик нажатия отдельной задачей - отменено

10:00 14.07.2023
  Активизация кнопок newBtn(), addBtn()
  Вывод sMODE, sHELP
Архив   MKlon3.5v7src_20230714.zip

  calcKeys()
        MKlon3.5v7src_20230714-1.zip

18:43 21.07.2023
На плате cpu отсутствует проводник CS на дисплей (X8/3 - SV1/8)

08:43 23.07.2023
board/button.h, .cpp закомментированы
Порты светодиодов: OUTPUT_OPEN_DRAIN заменил на OUTPUT - не выяснено, проблема с RED
Измерение температуры         - OK
Регулирование скорости        - OK 

08:23 30.07.2023
Отменены инициализации кнопок

18:23 06.08.2023
Эксперименты с очередью

08:00 07.08.2023
Версия проекта MKlon3.5v7-2 с использованием очереди      - отменено

21:00 17.08.2023
Артефакты на кнопках устранены приостановкой задачи displayTask на время
отрисовки инверсного и нормального отображения активированной кнопки.     - OK
Случайная смена шрифта режима и помощи устранена приостановкой задачи 
xTask_Main на время отрисовки длинных строк в главном окне                - OK

21:00 18.08.2023
Оптимизация выдачи на дисплей.

21:00 21.08.2023
BOOT отключен
Восстановление режима при подаче питания                  - OK

Режим DEVICE:
Ввод смещения, множителя и сглаживания по напряжению      - OK
Ввод смещения, множителя и сглаживания по току            -  write_offset_i ?
Ввод Kp, Ki, Keyboard                                     - OK
Ввод частоты ПИД-регулятора                               - OK
Калибровка тачскрина                                      - OK
Очистка ключей                                            - OK

Режим CCCV:
Старт, коррекция minV, maxV, minI, maxI                   - OK
Задержка включения                                        - OK
CC  -
CV  -

OPTIONS:     
Postpone    - OK
Timeout     - OK
NominalV    - OK
Capacity    - OK
Akb         - OK

mdisplay : 
Очистка поля строки в главном окне                          - OK
Очистка главного окна при выводе нулевой строки отменена    - OK

21:00 27.08.2023

src/main.cpp
/*
  project:      MKLON3.5v7
  pcb:          cpu3.5v7a  
  pcb:          pow3.5v7
  display:      480*320 IC: ILI94856 + XPT2046
  driver:       SAMD21 M0 MINI
  C/C++:        1.17.5
  VS:           1.81.1  1.82.2
  PIO:          v3.3.1
  Arduino       v0.6.230727001 Arduino for Visual Studio Code
  Espressif 32: 3.5.0 (с 4.0 не совместимо)
                6.3.1 не совместимо
  date:         20230905
*/    

#include "board/mboard.h"
#include "board/msupervisor.h"
#include "driver/mcommands.h"
#include "display/mdisplay.h"
#include "mtools.h"
#include "mdispatcher.h"
#include "mconnmng.h"
#include "measure/mmeasure.h"
#include "connect/connectfsm.h"
#include <Arduino.h>


// Идентификаторы задач
TaskHandle_t xTask_Display;     // Вывод текущих данных на дисплей
TaskHandle_t xTask_Cool;        // Управления системой теплоотвода
TaskHandle_t xTask_Main;        // Выбор режима работы и его исполнение
TaskHandle_t xTask_Measure;     // Измерения напряжения питания и температуры
TaskHandle_t xTask_Driver;      // Отправка команд драйверу SAMD21
TaskHandle_t xTask_Connect;     // Подключения к WiFi сети
TaskHandle_t xTask_Touch;       // Опрос активных кнопок

// // Идентификаторы семафоров    // В этом проекте не используется
// SemaphoreHandle_t tftMutex;    // Семафор вывода сообщений на дисплей

portMUX_TYPE timerMux = portMUX_INITIALIZER_UNLOCKED;

static MBoard      * Board      = 0;
static MDisplay    * Display    = 0;
static MTools      * Tools      = 0;
static MCommands   * Commands   = 0;
static MMeasure    * Measure    = 0;
static MDispatcher * Dispatcher = 0;
#ifndef NOWIFI
  static MConnect    * Connect    = 0;
#endif

#ifndef NOWIFI
  void connectTask ( void * );
#endif
void displayTask ( void * );
void coolTask    ( void * );
void mainTask    ( void * );
void measureTask ( void * );
void driverTask  ( void * );
void touchTask   ( void * );


void setup()
{
  Serial.begin(115200);
  while(!Serial) {}
  
  Display    = new MDisplay();
  Board      = new MBoard();
  Tools      = new MTools(Board, Display);
  Commands   = new MCommands(Tools);
  Measure    = new MMeasure(Tools);
  Dispatcher = new MDispatcher(Tools);
  #ifndef NOWIFI
    Connect    = new MConnect(Tools);
  #endif

  // tftMutex = xSemaphoreCreateMutex();     // В этом проекте не используется
  // if ( tftMutex == NULL )  Serial.println("Mutex can not be created");
  // else                     Serial.print("\nm : ");   Serial.println( (int) tftMutex );

  // Выделение ресурсов для каждой задачи: память, приоритет, ядро.
  // Все задачи исполняются ядром 1, ядро 0 выделено для радиочастотных задач - BT и WiFi.
  #ifndef NOWIFI
    xTaskCreatePinnedToCore( connectTask, "Connect", 10000, NULL, 1, &xTask_Connect, 0 ); 
  #endif
  xTaskCreatePinnedToCore( mainTask,    "Main",     4096, NULL, 3, &xTask_Main,    1 ); 
  xTaskCreatePinnedToCore( displayTask, "Display",  4096, NULL, 2, &xTask_Display, 1 ); 
  xTaskCreatePinnedToCore( touchTask,   "Touch",    4096, NULL, 1, &xTask_Touch,   1 ); 
  xTaskCreatePinnedToCore( coolTask,    "Cool",     2048, NULL, 1, &xTask_Cool,    1 ); 
  xTaskCreatePinnedToCore( measureTask, "Measure",  4096, NULL, 1, &xTask_Measure, 1 ); 
  xTaskCreatePinnedToCore( driverTask,  "Driver",   4096, NULL, 4, &xTask_Driver,  1 ); 
}

void loop() 
{
  if( Serial2.available() )    // В буфере приема есть принятые байты, не факт, что пакет полный
  {                            // Пока не принят весь пакет, время ожидания ограничено 5мс
    vTaskEnterCritical(&timerMux);
      Tools->setErr( Commands->dataProcessing() );
    vTaskExitCritical(&timerMux);
  }
}

// Задача подключения к WiFi сети (полностью заимствована как есть)
#ifndef NOWIFI
  void connectTask( void * )
  {
    while(true) {
    //unsigned long start = millis();   // Старт таймера 
    //Serial.print("*");
      Connect->run(); 
      // Период вызова задачи задается в TICK'ах, TICK по умолчанию равен 1мс.
      vTaskDelay( 10 / portTICK_PERIOD_MS );
    // Для удовлетворения любопытства о длительности выполнения задачи - раскомментировать как и таймер.
    //Serial.print("Autoconnect: Core "); Serial.print(xPortGetCoreID()); Serial.print(" Time = "); Serial.print(millis() - start); Serial.println(" mS");
    // Core 1, 2...3 mS
    }
    vTaskDelete( NULL );
  }
#endif

// Задача выдачи данных на дисплей
void displayTask( void *pvParameters )
{
  while(true)
  {
    const TickType_t xDelay = pdMS_TO_TICKS(125);
    Display->runDisplay( Board->getCelsius() );
    vTaskDelay( xDelay );
  }
  vTaskDelete( NULL );
}

// Задача управления системой теплоотвода. Предполагается расширить функциональность, добавив 
// слежение за правильностью подключения нагрузки, масштабирование тока и т.д. 
void coolTask( void * )
{
  while (true)
  {
    Board->Supervisor->runCool();
    vTaskDelay( 200 / portTICK_PERIOD_MS );
  }
  vTaskDelete( NULL );
}

// 1. Задача обслуживает выбор режима работы.
// 2. Управляет конечным автоматом выбранного режима вплоть да выхода из режима.
// И то и другое построены как конечные автоматы (FSM).
void mainTask ( void * )
{ 
  while (true)
  {
    // Выдерживается период запуска для вычисления амперчасов. Если прочие задачи исполняются в порядке 
    // очереди, то эта точно по таймеру - через 0,1с.
    portTickType xLastWakeTime = xTaskGetTickCount();   // To count the amp hours
    Dispatcher->run(); 
    vTaskDelayUntil( &xLastWakeTime, 100 / portTICK_PERIOD_MS );    // период 0,1с
  }
  vTaskDelete( NULL );
}

// Задача управления измерениями напряжения источника питания и датчика температуры
void measureTask( void * )
{
  while (true)
  {
    Measure->run();
    vTaskDelay( 100 / portTICK_PERIOD_MS );
  }
  vTaskDelete(NULL);
}

// Отправка команд драйверу SAMD21
void driverTask( void * )
{
  while (true)
  {
    vTaskEnterCritical(&timerMux);
    Commands->doCommand();
    vTaskExitCritical(&timerMux);
    vTaskDelay( 75 / portTICK_PERIOD_MS );
  }
  vTaskDelete(NULL);
}

// Задача опроса активных кнопок
void touchTask( void * )
{
  while(true)
  {
    Display->calcKeys();
    vTaskDelay( 50 / portTICK_PERIOD_MS );
  }
  vTaskDelete( NULL );
}
